compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp

```diff
@@ -528,6 +528,149 @@ wrapOpInDispatchRegion(RewriterBase &rewriter, Operation *op) {
+                                          Operation *op) {
+  assert(op && !isNonNullAndOutsideDispatch(op) &&
+         "op expected to be in a dispatch");
+  OpBuilder::InsertionGuard g(rewriter);
+  auto dispatchRegionOp = op->getParentOfType<DispatchRegionOp>();
+  // If all operands of the `op` come from outside the dispatch, then the op can
+  // be hoisted out before the dispatch region. Otherwise, the op can be hoisted
+  // out below the dispatch if the only users of the op are the dispatch return.
+  if (llvm::none_of(op->getOperands(), [&](Value operand) {
+        Operation *producer = operand.getDefiningOp();
+        return producer && producer->getParentOfType<DispatchRegionOp>();
+      })) {
+    rewriter.setInsertionPoint(dispatchRegionOp);
+  } else if (llvm::all_of(op->getUsers(), [&](Operation *user) {
+               return isa<IREE::Flow::ReturnOp>(user);
+             })) {
+    rewriter.setInsertionPointAfter(dispatchRegionOp);
+  } else {
+    return rewriter.notifyMatchFailure(
+        op, "op has both operands and users insided of its dispatch");
+  }
+  Operation *hoistedOp = rewriter.clone(*op);
+  rewriter.setInsertionPoint(dispatchRegionOp);
+
+  // Get the new dispatch region result types without the hoisted op results.
+  auto dispatchReturnOp = cast<IREE::Flow::ReturnOp>(
+      dispatchRegionOp.getBody().front().getTerminator());
+  SmallVector<Type> newResultTypes;
+  SmallVector<OpOperand *> newDispatchReturnOperands;
+  for (OpOperand &operand : dispatchReturnOp->getOpOperands()) {
+    if (operand.get().getDefiningOp() != op) {
+      newResultTypes.push_back(operand.get().getType());
+      newDispatchReturnOperands.push_back(&operand);
+    }
+  }
+
+  // Replace op uses inside and outside of the dispatch region with the hoisted
+  // results.
+  bool yieldsResults = false;
+  for (OpResult result : op->getResults()) {
+    Value hoistedResult = hoistedOp->getResult(result.getResultNumber());
+    // Replace all results yielded by the dispatch region with the hoisted
+    // op results.
+    for (OpOperand &use : result.getUses()) {
+      if (!isa<IREE::Flow::ReturnOp>(use.getOwner())) {
+        continue;
+      }
+      yieldsResults = true;
+      Value dispResult = dispatchRegionOp.getResults()[use.getOperandNumber()];
+      rewriter.replaceAllUsesWith(dispResult, hoistedResult);
+    }
+    // Replace uses inside the dispatch region.
+    rewriter.replaceAllUsesWith(result, hoistedResult);
+  }
+  // If no results were yielded from `op`, then nothing more to do.
+  if (!yieldsResults) {
+    return hoistedOp;
+  }
+
+  // Add the operands of the `op` to the new return values of the dispatch.
+  SmallVector<Value> newReturnedValues = llvm::map_to_vector(
+      newDispatchReturnOperands, [](OpOperand *val) { return val->get(); });
+  SetVector<Value> newReturnedValueSet(newReturnedValues.begin(),
+                                       newReturnedValues.end());
```

Comment:
I'm not sure if it works or not. Perhaps we can use `SmallSetVector`, then we don't need the `SetVector`. Can you give it a try?

https://llvm.org/doxygen/SetVector_8h_source.html